src/hardware/gus.cpp

```diff
@@ -628,44 +623,112 @@ void Gus::ActivateVoices(uint8_t requested_voices)
-
-		// Zero our buffer. The audio sequence for each active voice
-		// will be accumulated one at a time by the buffer's elements.
-		assert(frames <= render_buffer.size());
-		const auto num_samples = frames * 2;
-		std::fill_n(render_buffer.begin(), num_samples, 0.0f);
-
-		if (dac_enabled) {
-			auto voice = voices.begin();
-			const auto last_voice = voice + active_voices;
-			while (voice < last_voice && *voice) {
-				voice->get()->GenerateSamples(render_buffer,
-				                              ram, vol_scalars,
-				                              pan_scalars, frames);
-				++voice;
-			}
+	AudioFrame accumulator = {};
+
+	if (dac_enabled) {
+
+		auto voice = voices.begin();
+		const auto voice_end = voice + active_voices;
+
+		while (voice < voice_end && *voice) {
+			const auto voice_frame = voice->get()->RenderFrame(
+			        ram, vol_scalars, pan_scalars);
+
+			accumulator.left += voice_frame.left;
+			accumulator.right += voice_frame.right;
+
+			++voice;
 		}
-		soft_limiter.Process(render_buffer, frames, play_buffer);
-		audio_channel->AddSamples_s16(frames, play_buffer.data());
-		CheckVoiceIrq();
-		generated_frames += frames;
+		accumulator.left *= accumulator_scalar.left;
+		accumulator.right *= accumulator_scalar.right;
+	}
+	        CheckVoiceIrq();
+	        return accumulator;
+}
+
+bool Gus::RenderForMs(const double interval_ms)
+{
+	// How many frames fall within the given duration?
+	auto frames_to_render = static_cast<int>(interval_ms * frame_rate_per_ms);
+
+	// Capture our return state
+	const auto did_some_rendering = frames_to_render > 0;
+
+	// Render and queue the frames, which will be drained by the callback
+	while (frames_to_render-- > 0)
+		fifo.emplace(RenderFrame());
+
+	return did_some_rendering;
+}
+
+void Gus::RenderUpToNow()
+{
+	unused_for_ms = 0;
+	const auto now = PIC_FullIndex();
+
+	if (audio_channel->is_enabled) {
+		if (RenderForMs(now - last_render_time)) {
+			last_render_time = now;
+		}
+		return;
+	}
+
+	// Otherwise wake up the channel and mark the new last-update time.
+	// Subsequent renderings will get the new stream of frames.
+	audio_channel->Enable(true);
+	last_render_time = now;
+}
+
+double Gus::ConvertFramesToMs(const int frames) const
+{
+	return frames / frame_rate_per_ms;
+}
+
+void Gus::AudioCallback(uint16_t requested_frames)
+{
+	assert(audio_channel);
+	while (requested_frames && fifo.size()) {
+		audio_channel->AddSamples_sfloat(1, &fifo.front()[0]);
+		fifo.pop();
+		--requested_frames;
+	}
+
+	if (requested_frames) {
+		last_render_time += ConvertFramesToMs(requested_frames);
+		while (requested_frames--) {
+			const auto frame = RenderFrame();
+			audio_channel->AddSamples_sfloat(1, &frame[0]);
+		}
+	}
+	// Pause the card if it hasn't been written to for 3 seconds
+	constexpr uint16_t three_seconds_of_ticks = 3 * 1000;
+	if (++unused_for_ms > three_seconds_of_ticks) {
+		audio_channel->Enable(false);
+		LOG_MSG("sleeping GUS!");
```

Comment:
Nice descriptive variable names everywhere — I like it! 🎉

Reply:
Use same prefix here too? E.g. "GUS: Sleeping"

Reply:
(thats for catching it! Will remove this sleep log..just a check to see if it's working)